src/java.base/share/classes/jdk/internal/constant/MethodTypeDescImpl.java

```diff
@@ -105,18 +106,66 @@ public static MethodTypeDescImpl ofValidated(ClassDesc returnType, ClassDesc...
      * @jvms 4.3.3 Method Descriptors
      */
     public static MethodTypeDescImpl ofDescriptor(String descriptor) {
-        // Implicit null-check of descriptor
-        List<ClassDesc> ptypes = ConstantUtils.parseMethodDescriptor(descriptor);
-        int args = ptypes.size() - 1;
-        ClassDesc[] paramTypes = args > 0
-                ? ptypes.subList(1, args + 1).toArray(ConstantUtils.EMPTY_CLASSDESC)
-                : ConstantUtils.EMPTY_CLASSDESC;
-
-        MethodTypeDescImpl result = ofValidated(ptypes.get(0), paramTypes);
+        int rightBracket, len;
+        if (descriptor.charAt(0) != '('
+                || (rightBracket = (descriptor.charAt(1) == ')' ? 1 : descriptor.lastIndexOf(')'))) <= 0
+                || (len = descriptor.length() - rightBracket - 1) == 0
+                || (len != 1 && len != ConstantUtils.skipOverFieldSignature(descriptor, rightBracket + 1, descriptor.length()))
+        ) {
+            throw badMethodDescriptor(descriptor);
+        }
+        var returnType = ConstantUtils.resolveClassDesc(descriptor, rightBracket + 1, len);
+        var paramTypes = paramTypes(descriptor, 1, rightBracket);
+        var result = new MethodTypeDescImpl(returnType, paramTypes);
         result.cachedDescriptorString = descriptor;
         return result;
     }
 
+    private static ClassDesc[] paramTypes(String descriptor, int start, int end) {
+        if (start == end) {
+            return ConstantUtils.EMPTY_CLASSDESC;
+        }
+
+        /*
+         * If the length of the first 8 parameters is <= 256, save them in lengths to avoid secondary scanning
+         */
+        long lengths = 0;
+        int paramCount = 0;
+        for (int cur = start; cur < end; ) {
+            int len = ConstantUtils.skipOverFieldSignature(descriptor, cur, end);
+            if (len == 0) {
+                throw badMethodDescriptor(descriptor);
+            }
+            if (paramCount < 8) {
+                // use 0 to indicate that the current parameter length is greater than 256 and needs to be reparsed
+                lengths = (lengths << 8) | (len > 0xFF ? 0 : len);
+            }
+            paramCount++;
+            cur += len;
+        }
+
+        var paramTypes = new ClassDesc[paramCount];
+        int paramIndex = 0,
+            lengthsEnd = Math.min(paramCount, 8) - 1;
+        for (int cur = start; cur < end; ) {
+            int len = 0,
+                num = lengthsEnd - paramIndex;
+            if (num >= 0) {
+                int shift = num << 3;
+                len = (int) ((lengths & (0xFFL << shift)) >>> shift);
+            }
+            if (len == 0) {
+                len = ConstantUtils.skipOverFieldSignature(descriptor, cur, end);
```

Comment:
This can use a simpler version of skip that doesn't validate, like:
```java
int start = cur;
while (descriptor.charAt(cur) == '[') {
    cur++;
}
if (descriptor.charAt(cur++) == 'L') {
    cur = descriptor.indexOf(';', cur) + 1;
}
paramTypes[paramIndex++] = ConstantUtils.resolveClassDesc(descriptor, start, cur - start);
```

Reply:
Now this version merges the loops, and cannot handle this when paramIndex >= 8.

Reply:
Hmm? I mean this block can be a new method (the cur pointer moving part) and replace existing skipOver call that does extra validation

Reply:
I refactored the skipOverFieldSignature method and used the structure you mentioned above, so that the code is more concise.